core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -378,6 +385,26 @@ class ReplicaFetcherThread(name: String,
     result
   }
 
+  /*
+    Decides whether or not we should send a LeaderEpochRequest
+      by checking the broker message format and all topics' message format versions
+   */
+  private def shouldSendLeaderEpochRequest(partitions: Map[TopicPartition, Int]): Boolean = {
+    if (!brokerSupportsLeaderEpochRequest)
+      return false
+
+    // if all topics' message format do not support epoch requests, skip it
+    partitions.map { case (tp, _) => topicSupportsEpochRequest(tp) }.reduce(_ || _)
+  }
+
+  // protected for test mocking
+  protected def topicSupportsEpochRequest(tp: TopicPartition): Boolean = {
+    replicaMgr.getLog(tp) match {
+      case Some(log) => log.config.messageFormatVersion >= KAFKA_0_11_0_IV2
```

Comment:
There is a little bit of trickiness here in that the leader may have its message format version updated before the follower. However, the follower always accepts the message format of the leader. I'm wondering if we should skip the config check and send the OffsetsForLeaderEpoch request iff we have an epoch to query?